internal_docs/generate_release_changelog/main.go

```diff
@@ -0,0 +1,255 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/cheggaaa/pb/v3"
+	"github.com/google/go-github/v28/github"
+	"github.com/spf13/pflag"
+	"golang.org/x/oauth2"
+)
+
+const uncategorizedSectionTitle = "Uncategorized"
+
+func getCommitsInRange(ctx context.Context, ghClient *github.Client, gitHubOrg string, gitHubRepo string,
+	fromRevision string, toRevision string) ([]github.RepositoryCommit, error) {
+
+	// Make sure that the "from" revision is behind "to" and obtain the commits
+	comparison, _, err := ghClient.Repositories.CompareCommits(ctx, gitHubOrg, gitHubRepo, fromRevision, toRevision)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compare commits from %q and to %q: %s", fromRevision, toRevision, err)
+	}
+	if comparison.GetStatus() != "ahead" || comparison.GetBehindBy() != 0 {
+		return nil, fmt.Errorf("'from' revision (%s) is not behind 'to' revision (%s)", fromRevision, toRevision)
+	}
+	return comparison.Commits, nil
+}
+
+func findPullRequestsInCommits(ctx context.Context, ghClient *github.Client, gitHubOrg string, gitHubRepo string,
+	commits []github.RepositoryCommit, commitProcessHook func()) ([]*github.PullRequest, error) {
+	// Store pull requests in visiting order, preventing duplicates
+	prs := []*github.PullRequest{}
+	visitedPRs := map[int]struct{}{}
+	for _, commit := range commits {
+		commitProcessHook()
```

Comment:
I am not super happy about this. It seems rather artificial. The alternative I considered is returning a channel and running asynchronously, but I think that's a bit overkill.